lib/system/web/src/keyboard_engine.rs

```diff
@@ -0,0 +1,170 @@
+use std::collections::{HashMap, BTreeSet};
+use std::cell::RefCell;
+use std::rc::Rc;
+use wasm_bindgen::JsCast;
+use wasm_bindgen::prelude::*;
+use itertools::Itertools;
+
+enum Action {
+    KeyUp,
+    KeyDown
+}
+
+type Bindings = HashMap<String, Box<dyn FnMut()>>;
+
+#[wasm_bindgen]
+pub struct KeyboardEngine {
+    bindings:        Rc<RefCell<Bindings>>,
+    _element:         web_sys::Element,
+    _prevent_default: bool,
+    //keydown_cb:    Closure<dyn FnMut(web_sys::KeyboardEvent)>,
+    //keyup_cb:      Closure<dyn FnMut(web_sys::KeyboardEvent)>,
+    _state:          Rc<RefCell<BTreeSet<String>>>
+}
+
+impl KeyboardEngine {
+    pub fn new(element: &web_sys::Element, prevent_default: bool) -> Self {
+        let state = Rc::new(RefCell::new(BTreeSet::new()));
+        let bindings = Rc::new(RefCell::new(HashMap::new()));
+        //let (keydown_cb, keyup_cb) = KeyboardEngine::init(element, bindings.clone(), state.clone());
+        KeyboardEngine::init(element, bindings.clone(), state.clone(), prevent_default);
+
+        Self {
+            bindings,
+            _element: element.clone(),
+            _prevent_default: prevent_default,
+            _state: state,
+
+            //keydown_cb: keydown_cb,
+            //keyup_cb: keyup_cb,
+        }
+    }
+
+    pub fn from_tag_name(html_tag: &str, prevent_default: bool) -> Self {
+        let window = web_sys::window().expect("no global window exists");
+        let document = window.document().expect("should have document on window");
```

Comment:
Before landing on master all functions should handle errors in a nice way, without panics. Errors should never be written as strings - they need to be datatypes which can pretty print when needed :)